com.unity.netcode.gameobjects/Runtime/Messaging/MessagingSystem.cs

```diff
@@ -226,6 +259,66 @@ private bool CanReceive(ulong clientId, Type messageType, FastBufferReader messa
             return true;
         }
 
+        internal void ReorderMessage(int desiredOrder, uint targetHash)
+        {
+            if (desiredOrder < 0)
+            {
+                // invalid message re-ordering request. Ignore.
+                return;
+            }
+
+            if (desiredOrder < m_ReverseTypeMap.Length &&
+                XXHash.Hash32(m_ReverseTypeMap[desiredOrder].FullName) == targetHash)
+            {
+                // matching positions and hashes. All good.
+                return;
+            }
+
+            // Since the message at `desiredOrder` is not the expected one,
+            // insert an empty placeholder and move the messages down
+            var typesAsList = m_ReverseTypeMap.ToList();
+            typesAsList.Insert(desiredOrder, null);
+            m_ReverseTypeMap = typesAsList.ToArray();
+            var handlersAsList = m_MessageHandlers.ToList();
+            handlersAsList.Insert(desiredOrder, null);
+            m_MessageHandlers = handlersAsList.ToArray();
+
+            // we added a dummy message, bump the end up
+            m_HighMessageType++;
+
+            int position = desiredOrder;
```

Comment:
[2/2] This goes with the "skip over if not found" comment.

The technique to synchronize relies on the server enumerating the message *in order*. So, after receiving the n'th message from the server, we know that all messages in range [0,n] are either correct, or if they're missing on the client, the mapping will be empty for these missing position. So, we must only check from n, if we don't want to break the previous order. It also follows that any message the client has, but the server doesn't enumerate will be pushed to the end, where the server will not trigger them.

Why this way? The vast majority of the time this code will not trigger anyway, everything will just fall in the right place. In edge cases, we'll have different sets and it does function, at the cost of some empty entries. In very very edge cases, we'll have different sets *and* hash collisions. When we do, this technique guarantees a stable ordering. So messages will still fall in the right position because they started lexicographically ordered. 

That only leaves one remaining infrequent possibility: a mismatched message that happens to share a hash with an existing message that is just before it lexicographically. This case, we accept will not work.

But to get this far, we picked this algorithm. (sorry for the novel). I'll comment a bit in code, so the code remains understandable.

Reply:
Yes, please add some commentary about these assumptions (especially the "this will happen in order" aspect) to help future-us maintain the thing.